File: medium/coins.cpp
#include<iostream>
#include<vector>
using namespace std;

long long int coins(long long int n){
	if(n<12)
		return n;
	else{
		long long int n1,n2,n3;
		n1=n/2;
		n2=n/3;
		n3=n/4;
		n1=max(n1,coins(n1));
		n2=max(n2,coins(n2));
		n3=max(n3,coins(n3));
		if(n1>(n1+n2+n3)){
			return n;
		}	
		else
			return (n1+n2+n3);
	}


}
int main(){
	long long int n;
	while(cin>>n){
		cout<<coins(n)<<endl;
	}
	return 0;
}

File: medium/coins.py
mem={0:0};
def main(n):
	if n in mem:
		return mem[n];
	else:
		mem[n]=max(n,main(n/2)+main(n/3)+main(n/4));
		return mem[n];

while True:
	try:
		print main(int(raw_input()));
	except:
		break;